test/e2e/cache/cache_server_test.go

```diff
@@ -0,0 +1,207 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"reflect"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/kcp-dev/logicalcluster/v2"
+	"github.com/stretchr/testify/require"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apimachineryerrors "k8s.io/apimachinery/pkg/util/errors"
+	genericrequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	cacheclient "github.com/kcp-dev/kcp/pkg/cache/client"
+	cacheserver "github.com/kcp-dev/kcp/pkg/cache/server"
+	cacheopitons "github.com/kcp-dev/kcp/pkg/cache/server/options"
+	"github.com/kcp-dev/kcp/pkg/embeddedetcd"
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+)
+
+type testScenario struct {
+	name string
+	work func(ctx context.Context, t *testing.T, cacheClientRT *rest.Config, cluster logicalcluster.Name, gvr schema.GroupVersionResource)
+}
+
+// scenarios holds all test scenarios
+var scenarios = []testScenario{
+	{"TestUIDOverwrite", testUID},
+	// TODO: allow for setting CreationTimestamp for replicated objects
+	// TODO: changing spec doesn't increase the Generation of a replicated object
+	// TODO: deleting an object with finalizers immediately removes the obj
+	// TODO: a shard name is assigned to a replicated obj
+	// TODO: schema is not enforced
+	// TODO: spec and status can be updated at the same time
+}
+
+type fakeAPIExport struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              spec   `json:"spec,omitempty"`
+	Status            status `json:"status,omitempty"`
+}
+
+type spec struct {
+	Size int `json:"size,omitempty"`
+}
+
+type status struct {
+	Condition string `json:"condition,omitempty"`
+}
+
+// testUID checks if predefining an UID when the proper annotation are set works
+func testUID(ctx context.Context, t *testing.T, cacheClientRT *rest.Config, cluster logicalcluster.Name, gvr schema.GroupVersionResource) {
+	cacheDynamicClient, err := dynamic.NewClusterForConfig(cacheClientRT)
+	require.NoError(t, err)
+	mangoDB := newFakeAPIExport("mangodb")
+	mangoDB.UID = "3"
+	mangoDB.Generation = 1
+	mangoDB.Annotations = map[string]string{genericrequest.AnnotationKey: "amber", "kcp.dev/cluster": cluster.String()}
+	mangoDBRaw, err := toUnstructured(mangoDB)
+	require.NoError(t, err)
+
+	t.Logf("Create abmer/%s/mangodb on the cache server", cluster)
```

Comment:
nit: we usually do cluster|namespace/name

Reply:
for the cache server we need to take into account a shard name